### PR TITLE
complete reverse chunk stream at offset 0  despite pending tree stream

### PR DESCRIPTION
**Pull request by 2beaucoup**

### CODE REVIEW

File: banyan/tests/build_check.rs [hunk 1]
```diff
@@ -307,6 +307,23 @@ async fn stream_test_simple() -> anyhow::Result<()> {
...
+    let trees = stream::once(async move { tree }).chain(stream::pending());
+    let _ = forest
+        .stream_trees_chunked_reverse(EmptyQuery, trees, 0u64..=0, &|_| ())
+        .map_ok(move |chunk| stream::iter(chunk.data))
+        .take_while(|x| future::ready(x.is_ok()))
+        .filter_map(|x| future::ready(x.ok()))
+        .flatten()
+        .collect::<Vec<_>>()
+        .await;
+}
```
Comment (2beaucoup) [inline banyan/tests/build_check.rs:325]: this would otherwise not complete

File: banyan/src/forest/stream.rs [hunk 2]
```diff
@@ -148,9 +148,11 @@ impl<
         S: Stream<Item = Tree<T>> + Send + 'static,
     {
         let end_offset_ref = Arc::new(AtomicU64::new(*range.end()));
+        let end_offset_ref2 = end_offset_ref.clone();
         let forest = self.clone();
         trees
             .filter_map(move |tree| future::ready(tree.into_inner()))
+            .take_while(move |_| future::ready(end_offset_ref2.load(Ordering::SeqCst) > 0))
```
Comment (2beaucoup) [inline banyan/src/forest/stream.rs:155]: Wonder why this works at all as we need at least one more tree to progress.

**Comment by rklaehn** [comment]
Thanks a lot for the reproducer.

I think there is a bug in the fix. You should check for the offset range being empty as a termination criterium, not for the start being 0.

How about this: I kinda like it more despite the .boxed() because it has fewer atomic accesses and seems clearer.

rust```
                let range = *range.start()..=end_offset;
                // we are done
                if range.is_empty() {
                    return stream::empty().boxed()
                }```

**Comment by rklaehn** [comment]
None of my proposed improvements work...

**Comment by rklaehn** [comment]
OK, except for the fact that it needs a custom stream combinator from our utils crate, I like this one the best so far:

https://github.com/Actyx/banyan/commit/6a88d325d83b3f6794921b886a7690070c019105

**Comment by rklaehn** [comment]
Now, how about this one?

https://github.com/Actyx/banyan/commit/f8cc34a0efb05cf975707b1677fc220bb95ae458

**Comment by rklaehn** [comment]
OK, now how about this?

https://github.com/Actyx/banyan/pull/75

**Comment by rklaehn** [comment]
The forward methods have this comment, punting on terminating the stream. So technically they are correct.

```
    /// Note that this method has no way to know when the query is done. So ending this stream,
    /// if desired, will have to be done by the caller using e.g. `take_while(...)`.
```

But I guess we could modify that and guarantee that we terminate once a chunk overlaps the end of the range...

**Comment by 2beaucoup** [comment]
Changed `stream_trees_chunked` as well. Would it be much effort to cut a release and upgrade actyx?

**Review by rklaehn — APPROVED** [review]

**Pull request merged**